src/draw.cpp

```diff
@@ -409,12 +408,54 @@ void NCursesInterface::resume()
     init();
 }
 
-void NCursesInterface::print_job_graph(Job::Map const &jobs, int max_jobs) const
+void NCursesInterface::print_job_graph(Job::Map const &jobs, int max_jobs, size_t total_jobs) const
 {
-    addch('[');
-
-    int cnt = 0;
+    typedef std::pair<int /* host colour */, bool /*remote job */> BUCKET_KEY;
+    std::map<BUCKET_KEY, size_t /* count */> counts;
+    if(max_jobs < total_jobs) {
+        for (auto const j: jobs) {
+            if(!j.second->active)
+                continue;
+
+            int color = 0;
+            auto const h = j.second->getClient();
+            if (h)
+                color = h->getColor();
+
+            bool remote = !j.second->is_local;
+
+            auto entry = counts.emplace(std::make_pair(color, remote), 0);
+            counts[entry.first->first] = entry.first->second + 1;
+        }
+        //std::sort(counts.begin(), counts.end(), [](const &std::pair<BUCKET_KEY, size_t> lhs, const &std::pair<BUCKET_KEY, size_t> rhs) { return a.second < b.second; });
+
+        float remainder = 0.0;
+        float scale = max_jobs / float(total_jobs);
+        float d_allocated = 0.0;  // for debugging
+        size_t d_all = 0;
+        // scale the counts and track the remainder
+        for(auto e: counts) {
+            d_all += e.second;
+            float scaled_count = e.second * scale;
+            float rounded_count = std::floor(scaled_count);
+            d_allocated += rounded_count;
+            remainder += scaled_count - rounded_count;
+            counts[e.first] = size_t(rounded_count);
+        }
+        assert(((float(d_all) / total_jobs) - ((d_allocated + remainder) / max_jobs)) < 0.005);
+
+        // distribute the remainder amongst the buckets
+        auto bucket = counts.rbegin();
+        for(size_t r = std::ceil(remainder); r > 0; r--) {
```

Comment:
That was my first analysis but the overflow is *a lot*, you can run the simulation and when the load gets high the graph ends up with well over 3 additional slots. I generally terminate at this point since its obviously something bad happened.

Regarding how to handle remainders in general, I'm pretty ambivalent. I think both approaches have big issues. When the ratios are relatively even and slots are handed out to the lowest ratio then there's a bunch of edge cases, however when the ratios are very divergent and one dominates and slots are handed out to the highest ratio, then you end up completely hiding the present of the lower ratio jobs. Any solution here is going to be a compromise because you will be hiding jobs no matter what and any decision about hiding information will be imperfect.

Reply:
> That was my first analysis but the overflow is _a lot_, you can run the simulation and when the load gets high the graph ends up with well over 3 additional slots. I generally terminate at this point since its obviously something bad happened.
> 
> Regarding how to handle remainders in general, I'm pretty ambivalent. I think both approaches have big issues. When the ratios are relatively even and slots are handed out to the lowest ratio then there's a bunch of edge cases, however when the ratios are very divergent and one dominates and slots are handed out to the highest ratio, then you end up completely hiding the present of the lower ratio jobs. Any solution here is going to be a compromise because you will be hiding jobs no matter what and any decision about hiding information will be imperfect.

Sure. Whatever you do will always result in lost information since you are scaling down the number of jobs to a smaller value; I guess I would prefer if the number of shown jobs was at least intuitive (that is, a larger number of shown jobs means there are a larger number of actual jobs)